File: src/main/java/org/example/eshop/service/ShippingCalculatorService.java
package org.example.eshop.service;

import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class ShippingCalculatorService {
    private final ShopProperties shopProperties;

    public ShippingCalculatorService(ShopProperties shopProperties) {
        this.shopProperties = shopProperties;
    }

    /**
     * Calculates shipping cost for a given zone and weight.
     * @return The shipping cost, or null if zone is invalid or weight exceeds maximum bracket
     */
    public BigDecimal calculateShippingCost(String zone, int totalWeightGrams) {
        if (zone == null) return null;
        ShippingZone shippingZone = shopProperties.getShipping().getZones().get(zone.toLowerCase(Locale.ROOT));
        if (shippingZone == null) return null;

        ShippingBracket bracket = shippingZone.getBrackets().stream()
                .sorted(Comparator.comparingInt(ShippingBracket::getWeight))
                .filter(b -> b.getWeight() >= totalWeightGrams)
                .findFirst()
                .orElse(null);
        return bracket != null ? bracket.getCost() : null;
    }

    /**
     * Gets all available shipping zones.
     */
    public Map<String, String> getAvailableZones() {
        Map<String, ShippingZone> zones = shopProperties.getShipping().getZones();
        // Preserve iteration order of the underlying map when possible
        Map<String, String> result = new LinkedHashMap<>();
        zones.forEach((k, v) -> result.put(k, v != null ? v.getName() : null));
        return result;
    }

    /**
     * Gets the weight brackets for a specific zone.
     */
    public List<ShippingBracket> getZoneBrackets(String zone) {
        if (zone == null) return List.of();
        ShippingZone shippingZone = shopProperties.getShipping().getZones().get(zone.toLowerCase(Locale.ROOT));
        if (shippingZone == null || shippingZone.getBrackets() == null) return List.of();
        return shippingZone.getBrackets();
    }

    /**
     * Calculates shipping cost with detailed breakdown.
     */
    public ShippingCalculationResult calculateShippingWithDetails(String zone, int totalWeightGrams) {
        String normalized = zone == null ? null : zone.toLowerCase(Locale.ROOT);
        ShippingZone shippingZone = normalized == null ? null : shopProperties.getShipping().getZones().get(normalized);

        if (shippingZone == null) {
            return new ShippingCalculationResult(zone, null, totalWeightGrams, null, null,
                    "Invalid shipping zone: " + zone);
        }

        ShippingBracket bracket = shippingZone.getBrackets().stream()
                .sorted(Comparator.comparingInt(ShippingBracket::getWeight))
                .filter(b -> b.getWeight() >= totalWeightGrams)
                .findFirst()
                .orElse(null);

        if (bracket == null) {
            return new ShippingCalculationResult(zone, shippingZone.getName(), totalWeightGrams, null, null,
                    "Weight " + totalWeightGrams + "g exceeds maximum shipping weight for zone " + shippingZone.getName());
        }

        return new ShippingCalculationResult(zone, shippingZone.getName(), totalWeightGrams, bracket.getCost(), bracket, null);
    }
}

File: src/main/kotlin/org/example/eshop/service/ShippingCalculatorService.kt
package org.example.eshop.service

import org.springframework.stereotype.Service
import java.math.BigDecimal

/**
 * Service for shipping cost calculations.
 * Calculates shipping costs based on destination zone and total weight.
 */
@Service
class ShippingCalculatorService(
    private val shopProperties: ShopProperties
) {

    /**
     * Calculates shipping cost for a given zone and weight.
     *
     * @param zone The shipping zone (domestic, eu, row)
     * @param totalWeightGrams The total weight in grams
     * @return The shipping cost, or null if zone is invalid or weight exceeds maximum bracket
     */
    fun calculateShippingCost(zone: String, totalWeightGrams: Int): BigDecimal? {
        val shippingZone = shopProperties.shipping.zones[zone.lowercase()] ?: return null
        
        // Find the appropriate bracket for the weight
        // Brackets are ordered by weight, find the first bracket that can handle this weight
        val bracket = shippingZone.brackets
            .sortedBy { it.weight }
            .firstOrNull { it.weight >= totalWeightGrams }
            
        return bracket?.cost
    }

    /**
     * Gets all available shipping zones.
     *
     * @return Map of zone keys to zone names
     */
    fun getAvailableZones(): Map<String, String> {
        return shopProperties.shipping.zones.mapValues { it.value.name }
    }

    /**
     * Gets the weight brackets for a specific zone.
     *
     * @param zone The shipping zone
     * @return List of shipping brackets for the zone, or empty list if zone doesn't exist
     */
    fun getZoneBrackets(zone: String): List<ShippingBracket> {
        return shopProperties.shipping.zones[zone.lowercase()]?.brackets ?: emptyList()
    }

    /**
     * Calculates shipping cost with detailed breakdown.
     *
     * @param zone The shipping zone
     * @param totalWeightGrams The total weight in grams
     * @return Shipping calculation result with details
     */
    fun calculateShippingWithDetails(zone: String, totalWeightGrams: Int): ShippingCalculationResult {
        val normalizedZone = zone.lowercase()
        val shippingZone = shopProperties.shipping.zones[normalizedZone]
        
        if (shippingZone == null) {
            return ShippingCalculationResult(
                zone = zone,
                zoneName = null,
                weightGrams = totalWeightGrams,
                cost = null,
                bracket = null,
                error = "Invalid shipping zone: $zone"
            )
        }

        val bracket = shippingZone.brackets
            .sortedBy { it.weight }
            .firstOrNull { it.weight >= totalWeightGrams }

        if (bracket == null) {
            return ShippingCalculationResult(
                zone = zone,
                zoneName = shippingZone.name,
                weightGrams = totalWeightGrams,
                cost = null,
                bracket = null,
                error = "Weight ${totalWeightGrams}g exceeds maximum shipping weight for zone ${shippingZone.name}"
            )
        }

        return ShippingCalculationResult(
            zone = zone,
            zoneName = shippingZone.name,
            weightGrams = totalWeightGrams,
            cost = bracket.cost,
            bracket = bracket,
            error = null
        )
    }
}

/**
 * Result of shipping calculation with detailed information.
 */
data class ShippingCalculationResult(
    val zone: String,
    val zoneName: String?,
    val weightGrams: Int,
    val cost: BigDecimal?,
    val bracket: ShippingBracket?,
    val error: String?
) {
    val isValid: Boolean get() = error == null && cost != null
}